android-sdk/README.md

```diff
@@ -0,0 +1,224 @@
+        this.serverUrl = config.getProperty("auth-server-url");
+        this.clientId = config.getProperty("client_id");
+        this.audience = config.getProperty("audience");
+        this.grantType = config.getProperty("grant_type");
+        this.subjectTokenType = config.getProperty("subject_token_type");
+        this.requestedTokenType = config.getProperty("requested_token_type");
+        this.realm = config.getProperty("realm");
+        this.core = core;
+    }
+}
+```
+
+#### Registering a Service Module and the ServiceModuleRegistry
+The Core SDK will provide a service registry that modules can register themselves with as well as register their dependencies.  During build MobileCore will call bootstrap on services in the order of those with the least dependencies to those with the most dependencies.
+
+Example registration of KeyCloakService
+```
+public class KeyCloakService implements ServiceModule {
+  static {
+    ServiceModuleRegistry.registerServiceModule("keycloak", KeyCloakService.class, "http");
+  }
+  /*snip*/
+}
+```
+
+ServiceModuleRegistry.java
+```
+
+/**
+ * Registry of service modules.  It is responsible for maintaining a mapping of service modules to
+ * their types as well as the dependencies each module expects.
+ */
+public final class ServiceModuleRegistry {
+    private static final Map<String, Class<? extends ServiceModule>> serviceTypeMap = new HashMap<>();
+    private static final Map<String, String> dependencyMap = new HashMap<>();
+    /**
+     * Register the looking for a service module to a class.
+     *
+     * @param type the type of module to use
+     * @param moduleClass the class that implements a service module type
+     * @param dependsOn the other serviceTypes this Service Depends on.
+     */
+    public static void registerServiceModule(String type, Class<? extends ServiceModule> moduleClass, String... dependsOn) {
+        serviceTypeMap.put(type, moduleClass);
+        for (String dependency : dependsOn) {
+            dependencyMap.put(type, dependency);
+        }
+    }
+
+}
+```
+** Discussion Point ** I'm not a fan of static methods during testing.  Perhaps there should be a method in the ServiceModule interface which returns a list of services that the service depends on.  MobileCore could then read that list by invoking that method as part of bootstrapping.
+
+#### Referencing a Service
+An instance of the core bassed during bootstrap can be used to reference other services.  MobileCore is responsible for ensuring that ServiceModule.bootstrap is not called until all declared dependencies are bootstrapped.
+
+SyncService.java
+```
+public class SyncService implements ServiceModule {
+
+  KeyCloakService keycloakService;
+  /*snip*/
+  public void bootstrap(MobileCore core, ServiceConfiguration config) {
+     this.keycloakService = core.getService("keycloak", KeyCloakService.class);
+     this.networkService = core.getService("network", NetworkService.class);
```

Comment:
What `getService` is getting as parameters? 
Why users will use strings and class to get service? 

It seems that we mixing 2 different things - internal core functionalities like networking with separate service SDK like sync. 

I think it will be just better for core to have explicit list of internal functionalities like networking by:

```java
// Usage in any module like sync
core.getNetwork().get(...);

// Extend core in keycloak service - if keycloak service is not part of the core.
core.getNetwork().setInterceptor(keycloakInterceptor)
```

Reply:
Explicit is always better, but maybe we only need to define some Enum or List or group of constants that define the list of services registered in core. I don't know if it's good or bad approach to have all internal functionalities as getters. For example what if I don't want to use Network or Keycloak in my application, then what happen if I invoke that functionality? (not a rhetorical question!). I guess the `getService` is more flexible from this point of view.
```java
core.getService(Services.NETWORK)
core.getService(Services.KEYCLOAK)
core.getService(Services.SYNC)
```